sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/EnsureRequirements.scala

```diff
@@ -83,7 +83,25 @@ case class EnsureRequirements(conf: SQLConf) extends Rule[SparkPlan] {
         numPartitionsSet.headOption
       }
 
-      val targetNumPartitions = requiredNumPartitions.getOrElse(childrenNumPartitions.max)
+      // maxNumPostShufflePartitions is usually larger than numShufflePartitions,
+      // which causes some bucket map join lose efficacy after enabling adaptive execution.
+      // Please see SPARK-29655 for more details.
+      val expectedChildrenNumPartitions = if (conf.adaptiveExecutionEnabled) {
+        val withoutShuffleChildrenNumPartitions =
+          childrenIndexes.filterNot(children(_).isInstanceOf[ShuffleExchangeExec])
+            .map(children(_).outputPartitioning.numPartitions).toSet
+        if (withoutShuffleChildrenNumPartitions.nonEmpty &&
+          conf.maxNumPostShufflePartitions > conf.numShufflePartitions) {
+          math.min(math.max(withoutShuffleChildrenNumPartitions.max, conf.numShufflePartitions),
+            conf.maxNumPostShufflePartitions)
```

Comment:
this looks a bit puzzled.

is the purpose to not to shuffle maxNumPostShufflePartitions if maxNumPostShufflePartitions is larger than numShufflePartitions?

Reply:
Sorry for late reply @viirya. Yes. you are right.